save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt

```diff
@@ -93,98 +112,126 @@ class ExecutionController(private val executionService: ExecutionService,
-        try {
-            requireNotNull(projectService.findByNameAndOrganization(name, organization)).id!!.let {
-                testExecutionService.deleteTestExecutionWithProjectId(it)
-                agentStatusService.deleteAgentStatusWithProjectId(it)
-                agentService.deleteAgentWithProjectId(it)
+        return projectService.findWithPermissionByNameAndOrganization(
+            authentication,
+            name,
+            organization,
+            Permission.DELETE,
+            messageIfNotFound = "Could not find the project with name: $name and owner: ${organization.name} or related objects",
+        )
+            .mapNotNull { it.id!! }
+            .map { id ->
+                testExecutionService.deleteTestExecutionWithProjectId(id)
+                agentStatusService.deleteAgentStatusWithProjectId(id)
+                agentService.deleteAgentWithProjectId(id)
                 executionService.deleteExecutionByProjectNameAndProjectOrganization(name, organization)
+                ResponseEntity.ok().build<String>()
             }
-        } catch (e: IllegalArgumentException) {
-            log.warn("Could not find the project with name: $name and owner: ${organization.name} or related objects", e)
-            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Failed to delete executions for the following reason: ${e.message}")
-        }
-        return ResponseEntity.status(HttpStatus.OK).build()
     }
 
     /**
+     * FixMe: do we need to add preconditions that only executions from a single project can be deleted in a single query?
+     *
      * @param executionIds list of ids
+     * @param authentication
      * @return ResponseEntity
      * @throws ResponseStatusException
      */
     @PostMapping("/api/execution/delete")
-    fun deleteExecutionsByExecutionIds(@RequestParam executionIds: List<Long>): ResponseEntity<String>? {
-        try {
-            testExecutionService.deleteTestExecutionByExecutionIds(executionIds)
-            agentStatusService.deleteAgentStatusWithExecutionIds(executionIds)
-            agentService.deleteAgentByExecutionIds(executionIds)
-            executionService.deleteExecutionByIds(executionIds)
-        } catch (e: IllegalArgumentException) {
-            log.warn("Could not find the following executions: $executionIds or related objects", e)
-            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Failed to delete executions for the following reason: ${e.message}")
+    fun deleteExecutionsByExecutionIds(@RequestParam executionIds: List<Long>, authentication: Authentication): Mono<ResponseEntity<*>> = Flux.fromIterable(executionIds)
+        .map { it to executionService.findExecution(it) }
+        .filterAndInvoke({ (id, _) -> log.warn("Cannot delete execution id=$id because it's missing in the DB") }) { (_, execution) ->
+            execution.isPresent
+        }
+        .map { (_, execution) -> execution.get() }
```

Comment:
Quite a long chain to get projects, and may be it will be useful in the future to reuse this logic, maybe we can extract this logic into separate method?

Reply:
Yes, it's long, but for now it's the only method, where we accept a list of execution IDs. So at first we should filter out non-existent, then map the rest to projects. Now we use this method only to delete a single execution, and even for the future it would make sense to simplify logic by failing if some IDs don't exist or belong to different projects.